### ISSUE TITLE
Generic or specific process definitions?

### ISSUE THREAD

**Issue by m-mohr** [Open-EO/openeo-api#77]
This came up in #60 and was suggested by @edzer:

> What we should do is work with generic functions like min, max, mean, sum etc. and apply them to dimensions rather than write combination functions. Instead of time_min(obj) we would use map(obj, "time", min). Similar for aggregate: aggregate(obj, predicate, fun).

We need to discuss how we want to have our processes defined. I'll contribute some thoughts, pros/cons and examples later.

Google Earth Engine does both. They have the generic functions (e.g. filter combined with Date and Bounds classes) and many wrappers around them (e.g. filterDate, filterBounds).

**Comment by m-mohr** [comment]
**_Work in progress...._**

## Specific functions

### Examples

Examples are mostly based on API v0.0.2.

Process graph:

```
```

Process discovery:

```
[
  {
    "process_id":"filter",
    "description":"Filters image collections using a specific filter.",
    "args":{
      "imagery":{
        "description":"array of input collections with one element"
      },
      "filter": {
      }
    }
  },
  {
    "process_id":"filter_daterange",
    "description":"Drops observations from a collection that have been captured before a start or after a given end date.",
    "args":{
      "imagery":{
        "description":"array of input collections with one element"
      },
      "from":{
        "description":"start date"
      },
      "to":{
        "description":"end date"
      }
    }
  },
  {
    "process_id":"filter_bbox",
    "description":"Drops observations from a collection that are located outside of a given bounding box.",
    "args":{
      "imagery":{
        "description":"array of input collections with one element"
      },
      "left":{
        "description":"left boundary (longitude / easting)"
      },
      "right":{
        "description":"right boundary (longitude / easting)"
      },
      "top":{
        "description":"top boundary (latitude / northing)"
      },
      "bottom":{
        "description":"bottom boundary (latitude / northing)"
      },
      "srs":{
        "description":"spatial reference system of boundaries as proj4 or EPSG:12345 like string"
      }
    }
  },
  {
    "process_id":"min_time",
    "description":"Finds the minimum value of time series for all bands of the input dataset.",
    "args":{
      "imagery":{
        "description":"array of input collections with one element"
      }
    }
  }
]
```

## Generic functions

A big question here is how to document the arguments of processes properly. It seems much more complex than the other definition. "Sub"-functions as `min` can't be reused here. Handling of arguments is not clear. For example, the description of min is not precise. It needs to take an array of multiple arguments. How does the program know what to insert/use? Are "sub arguments" injected as set? Or as additional parameters? That will be hard to implement for languages like Java. ...

### Examples
Process graph:

```
```

Process discovery:

```
[
  {
    "process_id":"filter",
    "description":"Filters image collections using a specific filter.",
    "args":{
      "imagery":{
        "description":"array of input collections with one element"
      },
      "filter":{
        "anyOf":[
          {
            "value":"daterange",
            "description":"Drops observations from a collection that have been captured before a start or after a given end date.",
            "args":{
              "from":{
                "description":"start date"
              },
              "to":{
                "description":"end date"
              }
            }
          },
          {
            "value":"bbox",
            "description":"Drops observations from a collection that are located outside of a given bounding box.",
            "args":{
              "left":{
                "description":"left boundary (longitude / easting)"
              },
              "right":{
                "description":"right boundary (longitude / easting)"
              },
              "top":{
                "description":"top boundary (latitude / northing)"
              },
              "bottom":{
                "description":"bottom boundary (latitude / northing)"
              },
              "srs":{
                "description":"spatial reference system of boundaries as proj4 or EPSG:12345 like string"
              }
            }
          }
        ]
      }
    }
  },
  {
    "process_id":"aggregate",
    "description":"Aggregates values in an image collection.",
    "args":{
      "imagery":{
        "description":"array of input collections with one element"
      },
      "dimension":[
        {
          "value":"time",
          "description":"Time..."
        }
      ],
      "function":[
        {
          "value":"min",
          "description":"Calculates the minimum.",
          "args":{
            "...":{
              "description":"Multiple values to compare..."
            }
          }
        }
      ]
    }
  }
]
```

## Comparison

Argument | Specific | Generic
------------ | ------------- | -------------
Complexity of process discovery | 0 | 0
Process graph size | + | -
Learning curve / Usability | + | 0
Flexibility / Extensibility | 0 | +
Easy to document | + | -

**Comment by edzer** [comment]
Thanks, Matthias! @mkadunc came up with this idea during the workshop at VITO. I think functional programming is an extremely powerful concept. Having generics `filter`, `aggregate` and `reduce` (and maybe more?) will make everything easier, including implementing back-ends and doing UDFs. Let's take some time and see how these ideas are being put to work in GEE, databases, R, and so on.

**Comment by m-mohr** [comment]
@edzer I extended the post above a bit...

I disagree that it makes everything(!) easier. Especially process discovery (and comparison between back-ends) will be more complex. Back-end implementations will be easier in functional programming languages for sure, but in others it might be harder to accomplish this structure. But overall getting this structure running won't have a deep impact time-wise, I think. The scope of functions doesn't change much so the implementation work should be similar.

GEE implements filters, reducers etc mostly as classes, which is more the object-oriented style. This would require a change in the process graph definition (now: every object has to be a process, but "daterange" or "bands" is not a process itself). 
Which databases one could look at?

**Comment by mkadunc** [comment]
I agree that not everything is easier, here are some of my thoughts on the subject:

Specifying the API of generic `filter`, `aggregate`, `reduce` processes (that can work on any combination of dimensions and with any function used inside the loop) is definitely more difficult than specifying the API of a small set of `filter_by_<something>` and `min_by_<something>` functions. Implementing the above generic processes on the back-end to work properly using any combination of user-provided dimensions and predicate processes is definitely a difficult task, much more so than having a small set of predefined specific processes.

The problem with a small set of predefined specific processes is that it won't remain small much longer. The set will inevitably grow because the users will require expressing more complex algorithms and they will come up with new use-cases all the time, As the number of such processes grows, so will the effort of specifying and implementing them - each on its own will still be easier to define and implement than the generic alternatives, but the number will become hard to manage and maintain in a consistent manner.

If we start with generic processes, we put in extra effort at the beginning, but after we've implemented that we have a vastly more flexible and powerful syntax compared to implementing a couple of (most frequently used) filters and aggregators. Implementation of any subsequent wrappers becomes trivial (I think we should still have wrappers for most common cases, and to improve the learning curve for new users).

Regarding your proposal, I think we should strive to make the filters and aggregators into processes as well to have as much flexibility and reuse as possible (my understanding is that that process graphs are basically functions, correct me if i'm wrong) - the 'predicate' argument of filter is a function that takes a (subset of) image collection and returns a single boolean; an aggregator (function used in aggregate ops) is a process that takes an image collection that is an array (usually 1-dimensional) and returns a single value - usually scalar such as median (but it could also be vector, e.g. returning 5th, 50th and 95th percentiles). I believe the whole concept is pretty simple - the easiest way to understand it IMO is having a look at:
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter as a model for OpenEO `filter`
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map as a model for OpenEO `aggregate`

The difference between Javascript's arrays and OpenEO is only in the splitting of a collection into iterable items over which these functions operate (js arrays always iterate over the first dimension, openEO iteration dimension will be determined by the caller of the generic functions)

The difficult bit in this whole design would be defining the 'process' in a wide-enough way for it to be useful as a parameter of such higher-order functions (e.g. process would need to support returning scalar values and booleans given a specific projection of a subset of an image collection)

**Comment by jdries** [comment]
To me it seems like we can continue using the already defined specific functions, which would not prevent us from adding generic filter/aggregate functions add a later time. For me, functional programming only becomes useful when you can actually pass truly custom filtering and aggregating functions (UDF's) in the process graph, but then we need a language-agnostic way to define these functions.
